### PR TITLE
Add SLURM and NQSV examples

### PR DESCRIPTION
**Pull request by kathoef**
Wrapper script with configurations for JUWELS, HLRN-B and NESH. ~~Mistral currently has problems with Singularity and will be added at a later point.~~ Mistral supports only older e.g. CentOS 7 and Ubuntu 14 and 16 Docker container images (see troubleshooting section at https://www.dkrz.de/up/systems/mistral/singularity) and will be added at a later point?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,23 @@
 # SLURM-Singularity-bindings
```
Comment (willirath) [inline README.md]: As this is pretty general now, let's rename? (Maybe in a different PR?)

```suggestion
# Batch-Singularity-bindings
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,23 @@
...
-Wrapper scripts to enable host-system SLURM commands inside the container
+
+Wrapper scripts that enable host-system SLURM commands inside the container.
```
Comment (willirath) [inline README.md]:
```suggestion
Wrapper scripts that enable host-system batch scheduler commands inside the container.
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,23 @@
...
+Should be working especially with Red Hat (e.g. CentOS, ...) and Debian (e.g. Ubuntu, ...) Linux flavours.
+Tested for CentOS (7 and 8), Debian (8, 9 and 10) and Ubuntu (18.04 and 20.04) container images.
+Please note, currently only `sinfo`, `squeue`, `sbatch` and `scancel` are supported.
```
Comment (willirath) [inline README.md]:
```suggestion
Please note, that for SLURM, currently only `sinfo`, `squeue`, `sbatch` and `scancel` are supported.
```

File: bind_slurm.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+CONFIG=nesh.conf
+
+
+
+# Load bindings from config.
+
+source ${CONFIG}
+
+echo Will use: ${CONFIG}
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
# Get system-specific bindings to use here.
CONFIG=batch-bindings.conf
source ${CONFIG} || { echo ${CONFIG} not found. Make sure to link or create it!; exit; }
echo Will use: ${BATCH_CONFIG_NAME}
```

File: bind_slurm.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+# Check if singularity is available.
+
```
Comment (willirath) [inline bind_slurm.sh]: Delete additional whitespace? (For me, syntax highlighting makes things like this easy to read without additional space.)
```suggestion
```

File: bind_slurm.sh [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+echo Will execute: ${ORIGINAL_SINGULARITY_COMMAND}
+
+while [[ "$#" -gt 0 ]]; do
+  case $1 in
+    *.sif) IMAGE="$1" ;;
+    *) ;;
+  esac
+  shift
+done
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
# Get Singularity image path.
# We'll Preserve the original command and then parse all inputs until we find an image file.

ORIGINAL_SINGULARITY_COMMAND="$@"
echo Will execute: ${ORIGINAL_SINGULARITY_COMMAND}

while [[ "$#" -gt 0 ]]; do
  case $1 in
    *.sif) IMAGE="$1" ;;
    *) ;;
  esac
  shift
done
```

File: bind_slurm.sh [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+#getent passwd >> ${TEMPDIR}/etc_passwd
+#getent group >> ${TEMPDIR}/etc_group
+
+RTEMPDIR=$(basename $TEMPDIR) # get relative temp dir path
+
+MERGED_PASSWD_GROUP="\
+${RTEMPDIR}/etc_passwd:/etc/passwd,\
+${RTEMPDIR}/etc_group:/etc/group\
+"
+
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
echo Enable host SLURM user for: ${IMAGE}

# Prepare binding of system-specific SLURM.
#
# We'll filter system users and if there is a slurm user, we'll add it to the 
# /etc/passwd and /etc/group files used in the container.
# 
# This is done very similar to how singularity enables the human user inside the container:
# First, briefly start the container to get /etc/passwd and /etc/groups from with, then
# append users and groups to the files and bind the files when finally starting the container.

# create temporary directory that'll be deleted automatically once the script exists successfully
TEMPDIR=$(mktemp -d $(pwd)/tmp.XXXXXXXX)
echo Temporary directory: ${TEMPDIR}
trap "echo Deleting: ${TEMPDIR}; rm -rf ${TEMPDIR}" 0

# Extract user and group info from container and (mayby) append the slurm user
singularity exec ${IMAGE} cp -p /etc/passwd ${TEMPDIR}/etc_passwd
singularity exec ${IMAGE} cp -p /etc/group ${TEMPDIR}/etc_group
grep slurm /etc/passwd >> ${TEMPDIR}/etc_passwd
grep slurm /etc/group >> ${TEMPDIR}/etc_group

# Note that if system users are not managed in /etc/..., we could use getent:
# getent passwd | grep slurm >> ${TEMPDIR}/etc_passwd
# getent group | grep slurm >> ${TEMPDIR}/etc_group

# Create bind paths
RTEMPDIR=$(basename $TEMPDIR) # get relative temp dir path

MERGED_PASSWD_GROUP="\
${RTEMPDIR}/etc_passwd:/etc/passwd,\
${RTEMPDIR}/etc_group:/etc/group\
"

```

File: bind_slurm.sh [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+# Setup Singularity bind mount and (library) path environment.
+
+if [[ ! -z "$SLURM_PREPEND_PATH" ]]; then
+export SINGULARITYENV_PREPEND_PATH=${SLURM_PREPEND_PATH}${SINGULARITYENV_PREPEND_PATH:+:$SINGULARITYENV_PREPEND_PATH}
+echo SINGULARITYENV_PREPEND_PATH: "${SINGULARITYENV_PREPEND_PATH}"
+fi
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
if [[ ! -z "$BATCH_PREPEND_PATH" ]]; then
export SINGULARITYENV_PREPEND_PATH=${BATCH_PREPEND_PATH}${SINGULARITYENV_PREPEND_PATH:+:$SINGULARITYENV_PREPEND_PATH}
echo SINGULARITYENV_PREPEND_PATH: "${SINGULARITYENV_PREPEND_PATH}"
fi
```

File: bind_slurm.sh [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+if [[ ! -z "${SLURM_LD_LIBRARY_PATH}" ]]; then
+export SINGULARITYENV_LD_LIBRARY_PATH=${SLURM_LD_LIBRARY_PATH}${SINGULARITYENV_LD_LIBRARY_PATH:+:$SINGULARITYENV_LD_LIBRARY_PATH}
+echo SINGULARITYENV_LD_LIBRARY_PATH: "${SINGULARITYENV_LD_LIBRARY_PATH}"
+fi
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
if [[ ! -z "${BATCH_LD_LIBRARY_PATH}" ]]; then
export SINGULARITYENV_LD_LIBRARY_PATH=${BATCH_LD_LIBRARY_PATH}${SINGULARITYENV_LD_LIBRARY_PATH:+:$SINGULARITYENV_LD_LIBRARY_PATH}
echo SINGULARITYENV_LD_LIBRARY_PATH: "${SINGULARITYENV_LD_LIBRARY_PATH}"
fi
```

File: hlrnb.conf [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+CONFIG_SYSTEM_SPECS="\
+/var/run/munge/,\
+/cm/shared/batch/slurm/20.02.4/install/lib/slurm/,\
+/cm/shared/batch/etc/,\
+/usr/lib64/libmunge.so.2\
+"
+
+SLURM_PREPEND_PATH=/cm/shared/batch/bin
+SLURM_LD_LIBRARY_PATH=/cm/shared/batch/slurm/20.02.4/install/lib/slurm
```
Comment (willirath) [inline hlrnb.conf]:
```suggestion
BATCH_PREPEND_PATH=/cm/shared/batch/bin
BATCH_LD_LIBRARY_PATH=/cm/shared/batch/slurm/20.02.4/install/lib/slurm
```

File: hlrnb.conf [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+# Specify HLRN-Berlin bindings.
+
+CONFIG_SLURM_COMMANDS="\
+/cm/shared/batch/bin/sinfo,\
+/cm/shared/batch/bin/squeue,\
+/cm/shared/batch/bin/sbatch,\
+/cm/shared/batch/bin/scancel\
+"
+
+CONFIG_SYSTEM_SPECS="\
```
Comment (willirath) [inline hlrnb.conf]:
```suggestion
BATCH_SYSTEM_SPECS="\
```

File: hlrnb.conf [hunk 12]
```diff
@@ -0,0 +1,65 @@
+
+# Specify HLRN-Berlin bindings.
+
+CONFIG_SLURM_COMMANDS="\
```
Comment (willirath) [inline hlrnb.conf]:
```suggestion
BATCH_COMMANDS="\
```

File: bind_slurm.sh [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+SINGULARITY_BIND=${CONFIG_SLURM_COMMANDS},${CONFIG_SYSTEM_SPECS}
```
Comment (willirath) [inline bind_slurm.sh]:
```suggestion
SINGULARITY_BIND=${BATCH_COMMANDS},${BATCH_SYSTEM_SPECS}
```

File: juwels.conf [hunk 14]
```diff
@@ -0,0 +1,66 @@
+
+# Specify JUWELS bindings.
+
+CONFIG_SLURM_COMMANDS="\
```
Comment (willirath) [inline juwels.conf]:
```suggestion
BATCH_COMMANDS="\
```

File: juwels.conf [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+/usr/bin/sinfo,\
+/usr/bin/squeue,\
+/usr/bin/sbatch,\
+/usr/bin/scancel\
+"
+
+CONFIG_SYSTEM_SPECS="\
```
Comment (willirath) [inline juwels.conf]:
```suggestion
BATCH_SYSTEM_SPECS="\
```

File: juwels.conf [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+/etc/slurm,\
+/usr/lib64/libmunge.so.2,\
+/var/run/munge,\
+/usr/lib64/liblua-5.1.so,\
+/usr/share/lua/5.1,\
+/usr/lib64/lua/5.1,\
+/opt/jsc/slurm/etc/globres.json\
+"
+
+SLURM_LD_LIBRARY_PATH=/usr/lib64
```
Comment (willirath) [inline juwels.conf]:
```suggestion
BATCH_LD_LIBRARY_PATH=/usr/lib64
```

File: nesh.conf [hunk 17]
```diff
@@ -0,0 +1,56 @@
+
+# Specify Nesh bindings.
+
+CONFIG_SLURM_COMMANDS="\
```
Comment (willirath) [inline nesh.conf]:
```suggestion
BATCH_COMMANDS="\
```

File: nesh.conf [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+/opt/nec/nqsv/bin/qsub,\
+/opt/nec/nqsv/bin/qstat,\
+/opt/nec/nqsv/bin/qdel\
+"
+
+CONFIG_SYSTEM_SPECS="\
```
Comment (willirath) [inline nesh.conf]:
```suggestion
BATCH_SYSTEM_SPECS="\
```

File: nesh.conf [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+/etc/opt/nec/nqsv/api_client.conf\
+"
+
+SLURM_PREPEND_PATH=/opt/nec/nqsv/bin
```
Comment (willirath) [inline nesh.conf]:
```suggestion
BATCH_PREPEND_PATH=/opt/nec/nqsv/bin
```

File: hlrnb.conf [hunk 20]
```diff
@@ -0,0 +1,65 @@
+
+# Specify HLRN-Berlin bindings.
+
```
Comment (willirath) [inline hlrnb.conf:3]:
```suggestion
# Specify HLRN-Berlin bindings.

BATCH_CONFIG_NAME=HLRN-Berlin

```

File: juwels.conf [hunk 21]
```diff
@@ -0,0 +1,66 @@
+
+# Specify JUWELS bindings.
+
```
Comment (willirath) [inline juwels.conf:3]:
```suggestion
# Specify JUWELS bindings.

BATCH_CONFIG_NAME=JUWELS

```

File: nesh.conf [hunk 22]
```diff
@@ -0,0 +1,56 @@
+
+# Specify Nesh bindings.
+
```
Comment (willirath) [inline nesh.conf:3]:
```suggestion
# Specify Nesh bindings.

BATCH_CONFIG_NAME=Nesh
```

File: README.md [hunk 23]
```diff
@@ -1,2 +1,23 @@
...
+
+```shell
+$ singularity pull docker://centos:8
+$ ./bind_slurm.sh singularity shell --cleanenv centos_8.sif
```
Comment (willirath) [inline README.md]: Rename script to `bind_batch.sh`?
```suggestion
$ ./bind_batch.sh singularity shell --cleanenv centos_8.sif
```

File: README.md [hunk 24]
```diff
@@ -1,2 +1,23 @@
 # SLURM-Singularity-bindings
-Wrapper scripts to enable host-system SLURM commands inside the container
+
+Wrapper scripts that enable host-system SLURM commands inside the container.
+Should be working especially with Red Hat (e.g. CentOS, ...) and Debian (e.g. Ubuntu, ...) Linux flavours.
+Tested for CentOS (7 and 8), Debian (8, 9 and 10) and Ubuntu (18.04 and 20.04) container images.
+Please note, currently only `sinfo`, `squeue`, `sbatch` and `scancel` are supported.
+
```
Comment (willirath) [inline README.md:26]:
`````suggestion

Before being able to use the `bind_batch.sh`,  you'll have to enable the configuration for your system. Do this with (e.g for Juwels)
```shell
ln -sf juwels.conf batch-bindings.conf
```

`````

**Review by willirath — CHANGES_REQUESTED** [review]
Many small suggestions that all are along the following lines:

1. Drop slurm specifics also internally as this is also for Nesh now (and may help generalizing to other systems).

2. Make configuration changes explicit and robust against upstream updates by linking the config file.

3. Be a little more verbose in comments.

File: README.md (hunk 1 above)
Comment (kathoef) [reply README.md]: I'd say we better choose: "Batch scheduler Singularity bindings"?

File: bind_slurm.sh (hunk 7 above)
Comment (kathoef) [reply bind_slurm.sh]: I will fix some typos, but commit this for now! :)

File: bind_slurm.sh (hunk 8 above)
Comment (kathoef) [reply bind_slurm.sh]: I will go for `SCHEDULER_PREPEND_PATH` in a separate commit.

File: bind_slurm.sh (hunk 9 above)
Comment (kathoef) [reply bind_slurm.sh]: I will go for `SCHEDULER_LD_LIBRARY_PATH` in a separate commit. :)

File: hlrnb.conf (hunk 10 above)
Comment (kathoef) [reply hlrnb.conf]: See above!

File: hlrnb.conf (hunk 11 above)
Comment (kathoef) [reply hlrnb.conf]: Again. :)

File: hlrnb.conf (hunk 12 above)
Comment (kathoef) [reply hlrnb.conf]: See above. :)

File: bind_slurm.sh (hunk 13 above)
Comment (kathoef) [reply bind_slurm.sh]: See above.

File: juwels.conf (hunk 21 above)
Comment (kathoef) [reply juwels.conf:3]: I will instead implement a `readlink batch-bindings.conf` approach. :)

**Comment by kathoef** [comment]
I have added the (forgotten?) HLRN-Göttingen compatibility log file and (finally!) completed the README. I am aware that the NESH config does not work anymore (and HLRN-Göttingen announced today, that they have updated their SLURM installation, so that should be checked too), but I would like to have a look at these in separate PRs... can you re-review the status quo, @willirath?

**Comment by kathoef** [comment]
(Well, by the way, it could be that the JUWELS config also does not work anymore...)

File: README.md [hunk 25]
```diff
@@ -1,2 +1,51 @@
-# SLURM-Singularity-bindings
-Wrapper scripts to enable host-system SLURM commands inside the container
+# Batch scheduler bindings for Singularity containers
+
+This repository provides a wrapper script that adds HPC system batch scheduler functionality to Singularity containers. It was designed with container image portability in mind. Therefore, it does not require any target-system specific adaption to a container's `Dockerfile` and instead mounts the required system users, executables, libraries and configuration files into the container's file system at startup (see e.g. the [compatibility tests](./test_image_compatibility) for which only default container base images were used).
```
Comment (willirath) [inline README.md]: mounts --> bind mounts?

File: README.md [hunk 26]
```diff
@@ -1,2 +1,51 @@
...
+| HLRN-G<br>Redhat-7.8_3.10.0<br>SLURM | v3.2.1 | ---<br>CentOS: 7, 8<br>---<br>Amazon Linux: 1, 2 | Ubuntu: 10, 12, 14, 16, 18, 20<br>CentOS: 6<br>Debian: 6, 7, 8, 9, 10<br>--- |
+| Mistral<br>Redhat-6.10_2.6.32<br>SLURM | v3.6.1 | Ubuntu: 10, 12, 14, 16<br>CentOS: 6, 7<br>Debian: 6, 7, 8, 9<br>Amazon Linux: 1 | Ubuntu: 18, 20<br>CentOS: 8<br>Debian: 10<br>Amazon Linux: 2 |
+
+> Please note, for containers to be compatible with the wrapper script it might also be required that an application's executables and libraries are not installed in one of the file system locations that are specified in the HPC system-specific configuration files. Singularity bind mounting takes precedence over the container's original file system contents, which may result in a broken application when a container application is started with the wrapper. It should be noted that, in general, the portability of a container image can be improved considerably by installing target applications not in a default file system location, but in application-specific folders inside e.g. the container's root path (and then explicitly adding these to the `$PATH` environment variable).
```
Comment (willirath) [inline README.md]: Block-Quote? I'd wrap this in `*` or `_`. But this is more a matter of taste, I think.

**Review by willirath — CHANGES_REQUESTED** [review]
Only two minor suggestions. But these are not critical. Feel free to merge as is.

File: README.md (hunk 25 above)
Comment (kathoef) [reply README.md]: Yes, done.

File: README.md (hunk 26 above)
Comment (kathoef) [reply README.md]: I removed the blockquote styling. It might be best to just have this as normal text.

**Pull request merged**